pkg/agent/externalnode/external_node_controller.go

```diff
@@ -0,0 +1,756 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package externalnode
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+	"antrea.io/antrea/pkg/agent/util"
+	"antrea.io/antrea/pkg/apis/controlplane/v1beta2"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	enlister "antrea.io/antrea/pkg/client/listers/crd/v1alpha1"
+	binding "antrea.io/antrea/pkg/ovs/openflow"
+	"antrea.io/antrea/pkg/ovs/ovsconfig"
+	"antrea.io/antrea/pkg/ovs/ovsctl"
+	"antrea.io/antrea/pkg/util/channel"
+	"antrea.io/antrea/pkg/util/env"
+	"antrea.io/antrea/pkg/util/externalnode"
+	"antrea.io/antrea/pkg/util/ip"
+)
+
+const (
+	controllerName = "ExternalNodeController"
+	// How long to wait before retrying the processing of an ExternalNode change.
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+	// Disable resyncing.
+	resyncPeriod time.Duration = 0
+
+	ovsExternalIDUplinkName      = "uplink-name"
+	ovsExternalIDUplinkPort      = "uplink-port"
+	ovsExternalIDEntityName      = "entity-name"
+	ovsExternalIDEntityNamespace = "entity-namespace"
+	ovsExternalIDIPs             = "ips"
```

Comment:
A reminder for this one.

Reply:
The IPs are also used in `restart` cases. If we do not cache the IPs, it may cause the `InterfaceConfig` for the ExternalEntity which is loaded from OVSDB has no IPs, and the `reconcile` logic in network_policy_controller may failed to replay the OpenFlow entries on the egress rules because the IPs are not provided. This is because network_policy_controller and external_node_controller are different goroutines.

Reply:
Ok. I got confused why we specify the IPs in ExternalNode, but not discover them from interfaces? In the case of DHCP, users may not IPs when creating the ExternalNode.

Reply:
By now, we have validation on the existence of IP in Interfaces configuration when creating ExternalNode, although we support using both name and IPs in it. 

This additional check is from cloud VM, that the interface name is not known from the cloud API, but IP is always known. 

As for the case of DHCP, I didn't get your point why the user knows the interface name but does not know the IP.

Reply:
In the DHCP (and other auto-IPAM) case, users might not know the IP in advance, and the IP can change too (e.g. lease renew failed); interface name is static. I am not talking about public cloud VMs.

But I am fine to support only user specified IPs for now.